test/test_c10d.py

```diff
@@ -1800,6 +1800,88 @@ def test_sync_reduction(self):
 
+    @skip_if_not_nccl
+    @skip_if_not_multigpu
+    def test_arbitrary_forward_return_value(self):
+        """
+        Note: this test can be sped up by only running it on a CPU module
+        once DistributedDataParallel supports them.
+        """
+        store = c10d.FileStore(self.file.name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        class ForwardReturnValueModule(nn.Module):
+            def __init__(self):
+                super(ForwardReturnValueModule, self).__init__()
+                self.fc1 = nn.Linear(2, 10, bias=False)
+                self.fc2 = nn.Linear(10, 4, bias=False)
+                self.fc3 = nn.Linear(4, 4, bias=False)
+                self.relu = nn.ReLU()
+
+            def forward(self, x, fn, use_fc3=False):
+                x = self.relu(self.fc1(x))
+                x = self.relu(self.fc2(x))
+                if use_fc3:
+                    x = self.fc3(x)
+                return fn(F.softmax(x, dim=1))
+
+        device_id = gpus_for_rank(self.world_size)[self.rank][0]
+        model = DistributedDataParallel(
+            ForwardReturnValueModule().float().to(device_id),
+            device_ids=[device_id],
+            process_group=process_group,
+        )
+
+        batch_size = 4
+        criterion = nn.CrossEntropyLoss()
+        input = torch.rand([batch_size, 2], dtype=torch.float)
+        target = torch.LongTensor([random.randrange(4) for _ in range(batch_size)]).to(device_id)
+
+        # Always run "backward" to ensure the reducer is called by autograd.
+        # If we don't correctly capture the output tensors from the return value,
+        # the reducer won't see a hook for the unused parameter, and throw an error.
+        # The correct capture is what we're testing in this function.
+        def test(box, unbox):
+            output = model(input, fn=box)
+            loss = criterion(unbox(output), target)
+            loss.backward()
+
+        # Test with identity return value
+        test(
+            box=lambda x: x,
+            unbox=lambda x: x,
+        )
+
+        # Test with list return value
+        test(
+            box=lambda x: ["foo", "bar", x],
+            unbox=lambda x: x[2],
+        )
+
+        # Test with tuple return value
+        test(
+            box=lambda x: ("foo", "bar", x),
+            unbox=lambda x: x[2],
+        )
+
+        # Test with dict return value
+        test(
+            box=lambda x: {"foo": "bar", "tensor": x},
+            unbox=lambda x: x["tensor"],
+        )
+
+        # Test with list with dict return value
+        test(
+            box=lambda x: ["foo", "bar", {"tensor": x}],
+            unbox=lambda x: x[2]["tensor"],
+        )
+
+        # Test with dict with list return value
+        test(
+            box=lambda x: {"foo": "bar", "list": [0, 1, x]},
+            unbox=lambda x: x["list"][2],
```

Comment:
will it be necessary to add some tests that contains multiple tensors?

Reply:
Good point. I updated the test to return 2 output tensors and make the loss computation use the second one. Then manually confirmed that only passing the first output tensor to the reducer results in a crash (it assumes it won't receive a gradient for fc3 where in fact it does).